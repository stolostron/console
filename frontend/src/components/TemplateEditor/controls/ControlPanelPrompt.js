'use strict'

import React from 'react'
import PropTypes from 'prop-types'

class ControlPanelPrompt extends React.Component {
  static propTypes = {
    control: PropTypes.object,
    i18n: PropTypes.func,
  }

  render() {
    const { control } = this.props
    const { id, prompts = {} } = control
    const { type } = prompts
    switch (type) {
      case 'link':
        return <React.Fragment key={id}>{this.renderLink(control)}</React.Fragment>
    }
    return null
  }

  renderLink(control) {
    const { prompts } = control
    const { prompt, url, icon, id } = prompts
    const { i18n } = this.props
    const text = i18n(prompt)

    const createPopupWindow = () => {
      window.open(`${window.location.origin}${url}`, `${id}`)
    }

    return (
      <React.Fragment>
        <div className="creation-view-controls-add-value-container bottom-right">
          <div
            id={id}
            className="creation-view-controls-add-button"
            tabIndex="0"
            role={'button'}
            onClick={createPopupWindow}
            onKeyPress={createPopupWindow}
          >
            {text}
            {icon}
          </div>
        </div>
      </React.Fragment>
    )
  }
}

export default ControlPanelPrompt
