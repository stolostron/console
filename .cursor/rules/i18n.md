# Internationalization Rules for ACM Console

## Translation Patterns

### Hook Usage
- **REQUIRED**: Use acm-i18next library for translations (ESLint enforced)
- **DO NOT** import directly from 'react-i18next' - ESLint will error
- Import and use the translation hook consistently:
```typescript
import { useTranslation } from '../lib/acm-i18next'

export function MyComponent() {
  const { t } = useTranslation()
  return <div>{t('My translated text')}</div>
}
```

### Complex Translations
- **REQUIRED**: Use Trans component from acm-i18next (ESLint enforced)
- **DO NOT** import Trans from 'react-i18next' directly
```typescript
import { Trans } from '../lib/acm-i18next'

<Trans 
  i18nKey="<bold>Warning:</bold> This action cannot be undone."
  components={{ bold: <strong /> }}
/>
```

### Translation Keys
- Use descriptive, hierarchical keys
- Avoid generic keys like 'button', 'text', 'label'
- Use consistent naming patterns
- Group related translations logically

Good examples:
```typescript
t('cluster.create.title')
t('cluster.create.description')
t('cluster.status.ready')
t('cluster.status.notReady')
```

### String Externalization
- All user-facing strings must be externalized
- No hardcoded English text in components
- Use translation keys even for single words
- Include context in translation keys when needed

### Pluralization
- Use i18next pluralization features:
```typescript
t('cluster.count', { count: clusterCount })
// In translation file:
// "cluster.count_one": "{{count}} cluster"
// "cluster.count_other": "{{count}} clusters"
```

### Variable Interpolation
- Use proper variable interpolation:
```typescript
t('cluster.created.success', { clusterName: cluster.name })
// In translation file:
// "cluster.created.success": "Cluster {{clusterName}} created successfully"
```

### Translation Files
- Store all translations in `public/locales/en/translation.json`
- Use nested objects for organization
- Keep translation files well-structured
- Use consistent formatting and indentation

### Testing Translations
- Test that all translation keys exist
- Verify text rendering with different content lengths
- Test variable interpolation works correctly
- Use i18next-parser to extract keys

### Performance
- Avoid creating translation functions in render loops
- Use useMemo for complex translation computations
- Load translations efficiently
- Cache translation results when appropriate

### Accessibility
- Ensure translated text works with screen readers
- Consider text direction for RTL languages
- Test text overflow with longer translations
- Maintain proper semantic structure
