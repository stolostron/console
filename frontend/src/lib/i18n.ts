/* Copyright Contributors to the Open Cluster Management project */

/* istanbul ignore file */

import i18n from 'i18next'
// eslint-disable-next-line @typescript-eslint/no-restricted-imports
import { initReactI18next } from 'react-i18next'
import LanguageDetector from 'i18next-browser-languagedetector'
import HttpApi from 'i18next-http-backend'
import { supportedLanguages } from './supportedLanguages'

i18n
    // pass the i18n instance to react-i18next
    .use(initReactI18next)
    // detect user language
    // learn more: https://github.com/i18next/i18next-browser-languageDetector
    .use(LanguageDetector)
    // fetch json files
    // learn more: https://github.com/i18next/i18next-http-backend
    .use(HttpApi)
    // init i18next
    .init({
        backend: {
            loadPath: '/multicloud/locales/{{lng}}/{{ns}}.json',
        },
        compatibilityJSON: 'v3',
        fallbackLng: ['en'], // if language is not supported or string is missing, fallback to English
        keySeparator: false, // this repo will use single level json
        interpolation: {
            escapeValue: false, // react handles this already
        },
        defaultNS: 'translation', // the default file for strings when using useTranslation, etc
        nsSeparator: '~',
        supportedLngs: supportedLanguages, // only languages from this array will attempt to be loaded
        simplifyPluralSuffix: true,
    })

export default i18n
