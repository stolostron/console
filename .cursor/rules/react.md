# React Rules for ACM Console

## Component Patterns

### Functional Components
- Use functional components with hooks exclusively
- Implement proper prop interfaces with TypeScript
- Use React.memo for expensive components
- Follow established naming conventions (PascalCase)

### Hook Usage
- Use custom hooks for business logic separation
- Implement proper dependency arrays in useEffect
- Use useCallback for event handlers passed as props
- Leverage useMemo for expensive computations

### State Management
- Use Recoil atoms through useSharedAtoms() pattern
- Access atoms with useRecoilValue() for reading
- Use useRecoilValueGetter() for async operations
- Follow established atom naming conventions

Example (based on actual codebase patterns):
```typescript
/* Copyright Contributors to the Open Cluster Management project */
import { useMemo } from 'react'
import { useRecoilValue } from 'recoil'
import { useTranslation } from '../lib/acm-i18next'
import { useSharedAtoms } from '../shared-recoil'
import { AcmTable } from '../ui-components'

export function ClusterList() {
  const { t } = useTranslation()
  const { managedClustersState } = useSharedAtoms()
  const clusters = useRecoilValue(managedClustersState)
  
  const filteredClusters = useMemo(
    () => clusters.filter(cluster => cluster.status?.conditions),
    [clusters]
  )
  
  return (
    <AcmTable
      items={filteredClusters}
      columns={clusterColumns}
      keyFn={cluster => cluster.metadata.uid}
    />
  )
}
```

### Conditional Rendering
- Use logical AND (`&&`) for simple conditional rendering (ACM standard pattern):
  ```typescript
  {condition && <Component />}
  {props.additionalLabels && (<span><AcmLabels /></span>)}
  {expandable && itemCount > columnCount && (<AcmButton />)}
  ```
- Use ternary operators for either/or scenarios:
  ```typescript
  {props.isCompact ? <Popover /> : renderLabelGroup()}
  {showAll ? t('Show less') : t('Show all')}
  ```
- Early returns for major conditional logic:
  ```typescript
  if (props.hidden) return <></>
  if (!data) return <LoadingState />
  ```

### Performance
- Use `React.memo` for components that receive the same props frequently
- Use `useMemo` for expensive calculations
- Use `useCallback` for functions passed to child components
- Implement lazy loading for heavy components
- Avoid inline object/array creation in render
- Avoid using index as the key prop value
- Avoid using inline functions in JSX or TSX blocks
- Use proper key props for list items

### Error Boundaries
- Implement error boundaries for route-level components
- Use proper error state management
- Provide user-friendly error messages
- Log errors for debugging purposes

### Accessibility
- Use semantic HTML elements
- Implement proper ARIA attributes
- Ensure keyboard navigation works
- Test with screen readers

### Component Organization
- Keep components small and focused on a single responsibility
- Use PascalCase for component names and component files
- Use descriptive component names that indicate the component's purpose
- Avoid abbreviations in component names
- Separate presentation from business logic
- Use proper file structure (components/, ui-components/, routes/)
- Export components properly with named exports
- Avoid props drilling where props are only passed down to child components
- Do not mutate props or state - create a copy array or object if possible
