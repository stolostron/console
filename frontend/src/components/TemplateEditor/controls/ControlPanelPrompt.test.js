/* Copyright Contributors to the Open Cluster Management project */
'use strict'

import React from 'react'
import ControlPanelPrompt from './ControlPanelPrompt'
import { render } from '@testing-library/react'

import i18n from 'i18next'

const t = i18n.t.bind(i18n)

export const control = {
    name: 'creation.app.name',
    tooltip: 'tooltip.creation.app.name',
    controlData: [],
    id: 'name',
    type: 'text',
    prompts: {
        prompt: 'creation.ocp.cloud.add.connection',
        type: 'link',
        url: 'mylink',
        positionBottomRight: true,
        id: 'add-provider-connection',
    },
}
const fn = jest.fn()

describe('ControlPanelPrompt component', () => {
    it('renders as expected', () => {
        const Component = () => {
            return (
                <ControlPanelPrompt key={'key'} control={control} controlId={'controlId'} handleChange={fn} i18n={t} />
            )
        }

        const { asFragment, rerender } = render(<Component />)
        expect(asFragment()).toMatchSnapshot()
        control.prompts.type = ''
        rerender(<Component />)
        expect(asFragment()).toMatchSnapshot()
        control.prompts = undefined
        rerender(<Component />)
    })
})
