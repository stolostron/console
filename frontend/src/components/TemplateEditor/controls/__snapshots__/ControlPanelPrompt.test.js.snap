// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlPanelPrompt component renders as expected 1`] = `
<DocumentFragment>
  <div
    class="creation-view-controls-add-value-container bottom-right"
  >
    <div
      class="creation-view-controls-add-button"
      id="add-provider-connection"
      role="button"
      tabindex="0"
    />
  </div>
</DocumentFragment>
`;

exports[`ControlPanelPrompt component renders as expected 2`] = `<DocumentFragment />`;
